### PR TITLE
feat: add new runtimeVersion to manifest

### PR DESCRIPTION
**Pull request by danez**
#### Summary

The option `nodeVersion` now works differently and allows different styles of how the Node.js version is specified.
If it cannot be coerced to a Node.js version, the `undefined` is added to the manifest, leaving the choice of runtime to origin. We could also change this to have zisi add the default to the manifest, but I didn't do it because I feel its nicer to have the default version only in one place.

How it is supposed to work is that `@netlify/build` sets `nodeVersion` to `AWS_LAMBDA_JS_VERSION || buildUserNodeVersion` and based on this input we set a new field `runtimeVersion` in the manifest, or not if it is not a valid runtime version.
open-api reads the manifest and sends `runtimeVersion || runtime`.  In the origin we then detect if it is a runtime shortcut 'js' or a full runtime image name.

Ref: netlify/pod-compute#33
Ref: netlify/pod-compute#64

This would also allow the runtime being set by the deployment config API.

### CODE REVIEW

File: src/manifest.ts [hunk 1]
```diff
@@ -40,21 +41,23 @@ export const createManifest = async ({ functions, path }: { functions: FunctionR
 }
 
 const formatFunctionForManifest = ({
```
Comment (khendrikse) [inline src/manifest.ts:43]: Nit: I'm not sure what this function is really doing still, the only thing I see is that it resolves the path :/

**Review by khendrikse — DISMISSED** [review]

File: README.md [hunk 2]
```diff
@@ -155,14 +154,16 @@ The following properties are accepted:
   - _Type_: `string`
   - _Default value_: undefined
 
-  A name to use when displaying the function in the Netlify UI. Populates the `displayName` property in the functions manifest for the specified function.
+  A name to use when displaying the function in the Netlify UI. Populates the `displayName` property in the functions
```
Comment (eduardoboucas) [inline README.md:157]: What's up with these formatting changes? Makes the diff a bit hard to parse.

File: src/runtimes/node/bundlers/esbuild/bundler_target.ts [hunk 3]
```diff
@@ -35,20 +27,22 @@ const getModuleFormat = async (
   extension: string,
   configVersion?: string,
 ): Promise<{ includedFiles: string[]; moduleFormat: ModuleFormat }> => {
-  if (extension === ModuleFileExtension.MJS && featureFlags.zisi_pure_esm_mjs) {
+  if (featureFlags.zisi_pure_esm_mjs && extension === ModuleFileExtension.MJS) {
```
Comment (eduardoboucas) [inline src/runtimes/node/bundlers/esbuild/bundler_target.ts:30]: nit: Should this be moved to the `if` block below?

**Review by eduardoboucas — COMMENTED** [review]
Can you add a note outlining the breaking changes?

File: src/manifest.ts [hunk 4]
```diff
@@ -40,21 +41,23 @@ export const createManifest = async ({ functions, path }: { functions: FunctionR
...
+  bundler,
+  displayName,
+  generator,
   mainFile,
   name,
   path,
   runtime,
+  runtimeID,
```
[the remark this answers is not in the document]
Comment (danez) [reply src/manifest.ts:51]: Yes this was also annoying me.
`runtime` is `js`, `rs`, `go`
Where `runtimeID` now is the actual runtime`nodejs18.x`

Imho, `runtime` should probably called `language`, but I didn't want to change an existing field. Other solution would be instead of `runtimeID` add a field called `nodeVersion` where we simply add the node version that zisi received and use it in bitballoon/FO to decide on the runtime. Although that would not allow the deployment config API to set a runtime, but maybe if they set a node-version would be okay?

[the remark this answers is not in the document]
Comment (biruwon) [reply src/manifest.ts:51]: I am waiting on this for reviewing the rest of the PRs, as this field it's going to be change it makes the other PRs obsolete or in needs of a change. My two cents: I need to always check in code what these 2 are.

`language` and `runtimeID` seems good for me, as it kind of match the description for [AWS lambda runtimes](https://docs.aws.amazon.com/lambda/latest/dg/lambda-runtimes.html).

By [this comment](https://github.com/netlify/pod-compute/issues/64#issuecomment-1490060429), looks like we are going to make a PR in every repository touching runtimes, so maybe the effort of renaming them is not that big?

File: src/runtimes/node/bundlers/esbuild/bundler_target.ts (hunk 3 above)
Comment (danez) [reply src/runtimes/node/bundlers/esbuild/bundler_target.ts:30]: These are two different FF.

Comment (eduardoboucas) [reply src/runtimes/node/bundlers/esbuild/bundler_target.ts:30]: Ah, my bad!

**Comment by danez** [comment]
> Can you add a note outlining the breaking changes?

No breaking changes anymore. the `nodeVersion` field now supports a wider range of values, but still supports what was supported before.

File: vitest.config.ts [hunk 5]
```diff
@@ -5,7 +5,7 @@ export default defineConfig({
   test: {
     setupFiles: ['./tests/helpers/vitest_setup.ts'],
     include: ['tests/**/*.test.ts'],
-    testTimeout: 30_000,
+    testTimeout: 60_000,
```
Comment (danez) [inline vitest.config.ts:8]: Some windows test unrelated to this commit were timing out.

File: src/manifest.ts (hunk 1 above)
Comment (danez) [reply src/manifest.ts:43]: I guess yeah resolving the path and limiting what we output. We might not return everything from `FunctionResult`, but haven't checked.

File: src/manifest.ts (hunk 4 above)
[the remark this answers is not in the document]
Comment (danez) [reply src/manifest.ts]: I went for using one field for either language or the runtime name.

File: README.md [hunk 6]
```diff
@@ -128,16 +128,15 @@ The following properties are accepted:
 - `nodeVersion`
 
   - _Type_: `string`\
-  - _Default value_: `16.x`
+  - _Default value_: `16`
 
-  The version of Node.js to use as the compilation target. Possible values:
+  The version of Node.js to use as the compilation target and reported execution runtime. Any valid and supported
```
Comment (eduardoboucas) [inline README.md:133]: Curious what "reported" means in this context?

File: src/runtimes/runtime.ts [hunk 7]
```diff
@@ -39,14 +39,15 @@ export interface ZipFunctionResult {
...
+  generator?: string
   inputs?: string[]
   includedFiles?: string[]
   inSourceConfig?: ISCValues
   nativeNodeModules?: object
   nodeModulesWithDynamicImports?: string[]
   path: string
-  displayName?: string
-  generator?: string
+  runtimeName?: string
```
Comment (eduardoboucas) [inline src/runtimes/runtime.ts:50]: My understanding is that this now contains the version (e.g. `nodejs18.x`), so it's technically not just a runtime _name_ but also version?

**Review by eduardoboucas — APPROVED** [review]

**Review by khendrikse — COMMENTED** [review]
I'm not seeing specifically weird stuff, and would approve, but I see Eduardo is also looking at this, and I think his review could be more valuable right now ❤️

File: README.md (hunk 6 above)
Comment (danez) [reply README.md]: changed

File: src/runtimes/runtime.ts (hunk 7 above)
Comment (danez) [reply src/runtimes/runtime.ts:50]: Any idea what I should name it?

File: src/manifest.ts [hunk 8]
```diff
@@ -9,6 +9,7 @@ interface ManifestFunction {
   name: string
   path: string
   runtime: string
+  runtimeName?: string
```
Comment (eduardoboucas) [inline src/manifest.ts:12]: As discussed on Slack, we should probably rename this to `runtimeVersion`?

**Review by eduardoboucas — CHANGES_REQUESTED** [review]
